### ISSUE TITLE
Performance issue receiving large packets

### ISSUE THREAD

**Issue by ra1u** [aaugustin/websockets#809]
When receiving large (1MB and more) sized packets, performance drastically drops due to unnecessary memory copying.

Particular issue is with underlying [asyncio/streams.py](https://github.com/python/cpython/blob/3.8/Lib/asyncio/streams.py) and code in [`readexactly(self, n)`](https://github.com/python/cpython/blob/3.8/Lib/asyncio/streams.py#L717-L732) . 
However it seems that issue is tricky to remove in this streams library due to nature how streams work as they are design around concept, push when data arrives with  `feed_data` that works as central data queue.

Counting on fact, that  web sockets relays on `read_exactly` there should exist workaround.
Instead of receiving data trough `feed_data` on can feed buffer of known size using [asyncio.loop.sock_recv_into](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.sock_recv_into) .
However such approach requires minor plumbing in internals by circumventing private function [_ensure_fd_no_transport](https://github.com/python/cpython/blob/3.8/Lib/asyncio/selector_events.py#L237) .

My proof of concept test for receiving large packets show performance increase by approx 2x.

Are users and maintainers interested in such approach and has anyone has any idea how this can go wrong?

Thank, You.

**Comment by aaugustin** [comment]
"Circumventing private function _ensure_fd_no_transport" doesn't quite sound like "minor plumbing in internals" when supporting multiple Python & OS versions is a requirement ;-)

In my experience fighting against asyncio internals is a losing fight — at least it's a fight I started losing when asyncio was still called tulip and I never won since then.

I'd rather focus the efforts on completing the switch to a Sans-I/O core, then see if we can optimize the buffering there.

I think we're stuck with at least 4 data copies in practice: NIC/OS -> Python buffer -> unmasking -> decompression -> UTF-8 decoding. I'm surprised that you're getting an end-to-end speedup of x2 just by optimising the first part, because asyncio's buffering isn't too bad (I did some benchmarking in this area a couple years ago).
